src/cmd/lint.js

```diff
@@ -5,11 +5,59 @@ import {FileFilter} from './build';
 
 const log = createLogger(__filename);
 
+// Flow types
+
+export type LinterLogLevel = 'debug' | 'fatal';
+export type LinterOutputType = 'text' | 'json';
+
+export type LinterCreatorParams = {
+  config: {
+    logLevel: LinterLogLevel,
+    stack: boolean,
+    pretty?: boolean,
+    metadata?: boolean,
+    output?: LinterOutputType,
+    boring?: boolean,
+    selfHosted?: boolean,
+    shouldScanFile: (fileName: string) => boolean,
+    _: Array<string>,
+  },
+};
+
+export type Linter = {
+  run: () => Promise<void>,
+};
+
+export type LinterCreatorFn = (params: LinterCreatorParams) => Linter;
+
+
+export type LintCmdParams = {
+  sourceDir: string,
+  verbose?: boolean,
+  selfHosted?: boolean,
+  boring?: boolean,
+  output?: LinterOutputType,
+  metadata?: boolean,
+  pretty?: boolean,
+};
+
+export type LintCmdExtraParams = {
+  createLinter: LinterCreatorFn,
+  fileFilter: FileFilter,
+};
+
+// Module internals & exports
+
 export default function lint(
-    {verbose, sourceDir, selfHosted, boring, output,
-     metadata, pretty}: Object,
-    {createLinter=defaultLinterCreator, fileFilter=new FileFilter()}
-    : Object = {}): Promise<void> {
+  {
+    verbose, sourceDir, selfHosted, boring, output,
+    metadata, pretty,
+  }: LintCmdParams,
```

Comment:
I find it unfortunate that you can't declare destructured types inline, thus the need to declare a separate type object and repeat all the keys. Judging from [this thread](https://github.com/facebook/flow/issues/235), I don't think it will happen anytime soon though -- [destructured rebinding](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment#Assigning_to_new_variable_names) is the culprit!